Arm unittest refactor of Add and Conv2D test cases

AdrianLundell opened the pull request:
This is a showcase of multiple improvements which can be done across all tests:
1. Move pipeline definition from tests to a general (flexible) pipeline class
  - Define a default step of stages using e.g. TosaPipelineBI()
  - Add custom config or debug stages using helper functions s.a. .change_args(), .add_stage(), .dump_artifact() etc.
  - Run the full pipeline using .run()
2. Move towards a pure pytest approach to remove dependencies on unittest and parametrize
3. Separate tests running on FVP from tests not running on FVP rather than configuring this from the command line
  - FVP tests are skipped if not installed
  - To filter out tests one may instead use pytest markers/name filtering
  - This should give a clearer picture of what has been tested
4. Introduces one favored way of marking tests as xfails, in the parameterize decorator

Erik-Lundell committed:
Arm unittest refactor of Add and Conv test cases

This is a showcase of multiple improvements which can be done across all tests:
1. Move pipeline definition from tests to a general (flexible) pipeline class
  - Define a default step of stages using e.g. TosaPipelineBI()
  - Add custom config or debug stages using helper functions s.a. .change_args(), .add_stage(), .dump() etc.
  - Run the full pipeline using .run()
2. Move towards a pure pytest approach to remove dependencies on unittest and parametrize
3. Separate tests running on FVP from tests not running on FVP rather than configuring this from the command line
  - FVP tests are skipped if not installed
  - To filter out tests one may instead use pytest markers/name filtering
  - This should give a clearer picture of what has been tested
4. Introduces one favored way of marking tests as xfails, in the parameterize decorator

Change-Id: Ieb88209fd7bcbb4b700bb970d6877ca6785752a4

AdrianLundell commented:
@digantdesai Curious to see what you think about this, in particular this removes unittest support in favour of a simpler pytest-only setup + it introduces one extra layer of complexity by wrapping test stages in the new test_pipeline.

AdrianLundell committed:
Merge branch 'pytorch:main' into dev/refactor_unittests

digantdesai commented on backends/arm/test/common.py:
```diff
@@ -185,3 +189,41 @@ def get_target_board(compile_spec: list[CompileSpec]) -> str | None:
     return None
+
+
+u55_fvp_mark = pytest.mark.skipif(
```
Nit: Make it more intuitive
```suggestion
SkipIfNoCS300 = pytest.mark.skipif(
```

AdrianLundell replied:
Done

digantdesai commented on backends/arm/test/common.py:
```diff
@@ -185,3 +189,41 @@ def get_target_board(compile_spec: list[CompileSpec]) -> str | None:
+):
+    """
+    Custom version of pytest.mark.parametrize with some syntatic sugar and added xfail functionality
+        - test_data is expected as a dict of (id, test_data) pairs
+        - alllows to specifiy a dict of (id, failure_reason) pairs to mark specific tests as xfail
```
If these two dicts are meant to be looked up using same keys then, perhaps,
```
test_data: dict[str, Tuple[Any, str]] 
```
where key is a string, value is a tuple(any_test_data, failure_reason)

digantdesai replied:
If you want you can take it even more fine grained and allow individual test_data entry failure reason.

AdrianLundell replied:
I think I prefer this way, what we often want to do is to reuse the same test data for multiple targets where only one is failing. Here we avoid having to create separate dicts for this, by specifying a unique xfail-dict just for one target.

digantdesai commented on backends/arm/test/common.py:
```diff
@@ -185,3 +189,41 @@ def get_target_board(compile_spec: list[CompileSpec]) -> str | None:
+    if xfails is None:
+        xfails = {}
+
+    def decorator_func(func):
```
Add a docstring.

AdrianLundell replied:
I didn't understand exactly where this docstring should be added, there is one right above. Should it be moved?

digantdesai replied:
I was thinking about adding another about decoror meta fn and what its doing. Just so make it more readable and makes the intent clear.

digantdesai commented on backends/arm/test/tester/test_pipeline.py:
```diff
@@ -0,0 +1,361 @@
+""" Generic type used for test data in the pipeline. Depends on which type the operator expects."""
+
+
+class BasePipeline(Generic[T]):
```
Nit: To help improve readability

```suggestion
class BasePipelineMaker(Generic[T]):
```

if you are not a fan for `BasePipelineFactory` or something along those lines :)

digantdesai commented on backends/arm/test/tester/test_pipeline.py:
```diff
@@ -0,0 +1,361 @@
+        )
+        self.add_stage(-1, self.tester.to_executorch)
+
+    def add_stage(self, pos: int, func: Callable, *args, **kwargs):
```
Add logging, this meta debugging can get hard. 🙏

digantdesai commented on backends/arm/test/tester/test_pipeline.py:
```diff
@@ -0,0 +1,361 @@
+
+
+class TosaPipelineBI(BasePipeline, Generic[T]):
+    def __init__(
```
doc string

digantdesai commented on backends/arm/test/tester/test_pipeline.py:
```diff
@@ -0,0 +1,361 @@
+
+
+class TosaPipelineMI(BasePipeline, Generic[T]):
+    def __init__(
```
docstring for the class

digantdesai commented on backends/arm/test/tester/test_pipeline.py:
```diff
@@ -0,0 +1,361 @@
+        self,
+        module: torch.nn.Module,
+        test_data: T,
+        aten_ops: str | List[str],
+        exir_ops: str | List[str],
```
can this be looked up in a table? OK if you want to pass it in

AdrianLundell replied:
I want to pass it in to allow to use these pipelines for any type of module, including networks and so on. The operators expected in a graph is also a type of test data so it seems reasonable to keep it in the tests.

digantdesai replied:
I meant given (almost always) 1:1 mapping between these two sets of operators, ignoring dim_order ones, but not a strong preference TBH.

digantdesai commented on backends/arm/test/ops/test_add.py:
```diff
@@ -1,178 +1,144 @@
+    "4d_ones_1": (torch.ones(1, 1, 4, 4),),
+    "4d_ones_2": (torch.ones(1, 3, 4, 2),),
+}
+T1 = Tuple[torch.Tensor]
```
seems more pain than worth IMHO, what are your thoughts on these typing hints?

AdrianLundell replied:
I kind of agree but it also seems kind of nice to have the quick feedback if you give wrong test data.

I have restructured the Add test case a bit, with that as a starting point I do not think it would be too annoying to follow this pattern.

digantdesai replied:
> nice to have the quick feedback if you give wrong test data
Yeah, I can see the value now.

digantdesai commented on backends/arm/test/ops/test_conv2d.py:
```diff
@@ -228,118 +231,72 @@ def forward(self, x):
+    pipeline.run()
+
+
+xfails = {
```
these should live close to these keys in the modules dict IMO

digantdesai reviewed (commented):
I love the new approach. It's easier to add tests, cleaner, and reduces code duplication.

However, I have a concern that this might lead to "blind copy-paste" of new tests without thoroughly thinking about the test cases, especially from new users. Over time, we may end up with solid coverage for things that are already covered in the pipeline generator, but leave gaps elsewhere. For example, consider the conv test from your PR. When adding a module with two_conv, if I were to manually add a test case, in edge IR after to_edge, I would check for two convolution nodes to be present. Although it's possible to add a stage in the pipeline object, someone more likely to just do what you did.

Originally, the tester was designed to address code duplication and redundancy, without taking away too much control. 

That said, at this point, my concerns might not be too high, so I think we can proceed with this approach. We can evaluate when to use this more vigilantly.

AdrianLundell committed:
Address PR comments

- Rename FVP markers
- Improve structure in Add test
- Add doc strings in test_pipelines
- Remove unused code in test_pipelines
- Additional logging in test_pipelines

Change-Id: I94e21d14c2132481249d88a51df5d4415dd348a4

AdrianLundell committed:
Merge branch 'main' of https://github.com/pytorch/executorch into dev/refactor_unittests

digantdesai commented:
LGTM, bunch of CI issues, not sure how many are related. Please make it green. Let me pull it in internally, see if something goes wrong with `parametrize`

AdrianLundell committed:
Merge branch 'pytorch:main' into dev/refactor_unittests

AdrianLundell commented:
Great, I'll look into it when CI is stable again

AdrianLundell committed:
Merge branch 'main' into dev/refactor_unittests

AdrianLundell committed:
Add check for executor runner + minor documentation fixes

Change-Id: I7cc0d06321282824d1af55f027cd15de0ce95b99

AdrianLundell committed:
Merge branch 'main' of https://github.com/pytorch/executorch into dev/refactor_unittests

AdrianLundell commented:
None of these CI-failures look related to me

AdrianLundell committed:
Merge branch 'main' of https://github.com/pytorch/executorch into dev/refactor_unittests

digantdesai commented:
rebase please, couldn't import it internally. Thanks.

AdrianLundell committed:
Merge branch 'main' of https://github.com/pytorch/executorch into dev/refactor_unittests

AdrianLundell committed:
Merge branch 'main' of https://github.com/pytorch/executorch into dev/refactor_unittests

AdrianLundell committed:
Merge branch 'main' of https://github.com/pytorch/executorch into dev/refactor_unittests

AdrianLundell committed:
Merge branch 'main' into dev/refactor_unittests

AdrianLundell commented:
@digantdesai Are you able to import it now?

digantdesai commented:
Trying again now.

zingo commented:
@digantdesai  will you handle the merged or should we do it ? Not really sure about best action when PR is "imported". :)
